test/extensions/quic_listeners/quiche/envoy_quic_writer_test.cc

```diff
@@ -0,0 +1,138 @@
+#include <memory>
+#include <string>
+
+#include "common/network/io_socket_error_impl.h"
+
+#include "extensions/quic_listeners/quiche/envoy_quic_packet_writer.h"
+
+#include "test/mocks/network/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Quic {
+
+class EnvoyQuicWriterTest : public ::testing::Test {
+public:
+  EnvoyQuicWriterTest() : envoy_quic_writer_(udp_listener_) {
+    self_address_.FromString("0.0.0.0");
+    quic::QuicIpAddress peer_ip;
+    peer_ip.FromString("127.0.0.1");
+    peer_address_ = quic::QuicSocketAddress(peer_ip, /*port=*/123);
+    EXPECT_CALL(udp_listener_, onDestroy());
+    ON_CALL(udp_listener_, send(_))
+        .WillByDefault(testing::Invoke([](const Network::UdpSendData& send_data) {
+          return Api::IoCallUint64Result(
+              send_data.buffer_.length(),
+              Api::IoErrorPtr(nullptr, Network::IoSocketError::deleteIoError));
+        }));
+  }
+
+protected:
+  testing::NiceMock<Network::MockUdpListener> udp_listener_;
+  quic::QuicIpAddress self_address_;
+  quic::QuicSocketAddress peer_address_;
+  EnvoyQuicPacketWriter envoy_quic_writer_;
+};
+
+TEST_F(EnvoyQuicWriterTest, AssertOnNonNullPacketOption) {
+  std::string str("Hello World!");
+  EXPECT_DEBUG_DEATH(envoy_quic_writer_.WritePacket(str.data(), str.length(), self_address_,
+                                                    peer_address_,
+                                                    reinterpret_cast<quic::PerPacketOptions*>(0x1)),
+                     "Per packet option is not supported yet.");
+}
+
+TEST_F(EnvoyQuicWriterTest, SendSuccessfully) {
+  std::string str("Hello World!");
+  EXPECT_CALL(udp_listener_, send(_))
+      .WillOnce(testing::Invoke([this, str](const Network::UdpSendData& send_data) {
+        EXPECT_EQ(peer_address_.ToString(), send_data.peer_address_->asString());
+        EXPECT_EQ(self_address_.ToString(), send_data.local_ip_->addressAsString());
+        EXPECT_EQ(str, send_data.buffer_.toString());
+        return Api::IoCallUint64Result(
+            str.length(), Api::IoErrorPtr(nullptr, Network::IoSocketError::deleteIoError));
+      }));
```

Comment:
nit: Consider put the bulk of the EXPECT_CALL into a function to make it easier to read. Your call.